spec.html

```diff
@@ -46724,6 +46897,83 @@ <h1>Proxy.revocable ( _target_, _handler_ )</h1>
     </emu-clause>
   </emu-clause>
 
+  <emu-clause id="sec-module-source-objects">
+    <h1>Module Source Objects</h1>
+    <p>Module Source Objects represent modules in their source import phase, which are not linked, instantiated or executed.</p>
+    <p>All Module Source Objects should have a prototype of %AbstractModuleSource%.prototype and must define a [[ModuleSourceRecord]] internal slot.</p>
+    <p>Hosts may define their own Module Source subclasses for custom module record types.</p>
+
+    <emu-clause id="sec-%abstractmodulesource%-constructor">
+      <h1>The %AbstractModuleSource% Constructor</h1>
+      <p>The %AbstractModuleSource% constructor:</p>
+      <ul>
+        <li>is <dfn>%AbstractModuleSource%</dfn>.</li>
+        <li>along with its corresponding prototype object, provides common properties that are inherited by module source constructors and their instances.</li>
+        <li>does not have a global name or appear as a property of the global object.</li>
+        <li>acts as the abstract superclass of the _ModuleSource_ constructor.</li>
+        <li>will throw an error when invoked, because it is an abstract class constructor. The module source constructors do not perform a `super` call to it.</li>
```

Comment:
i'm not sure we have the concept of an "abstract" class constructor; can't it be a normal constructor that throws both when [[Call]]ed or when it's its own `new.target`?

Reply:
@ljharb Do you mean you want it to be possible to do like

```js
let AbstractModuleSource = /* some coe which gets AbstractModuleSource somehow */;
class A extends AbstractModuleSource {}

new A()
```
and have that not throw?

If so, that can't be done while keeping the property mentioned above, that these things are supposed to all have a [[ModuleSourceRecord]] internal slot holding a ModuleSource Record - there's no way to provide one when constructing from user code.

---

Or do you just mean you don't like the wording "abstract class constructor" here, and are OK with the behavior but want different wording?

Reply:
The semantics are fine, i indeed just mean the word "abstract" doesn't appear in the spec in this context afaik, and I don't think we should introduce it short of a proposal to add some kind of language feature for that.

Reply:
> can't it be a normal constructor that throws both when [[Call]]ed or when it's its own `new.target`?

Yes, we should do this, and we should get rid of 

> The module source constructors do not perform a `super` call to it

Reply:
What is the resolution of this? The current version I'm reading has an %AbstractModuleSource% constructor that unconditionally throws for both [[Call]] and [[Construct]].

Reply:
User code should be able to `class extends AbstractModuleSource`, though, otherwise you can't self-host a JS implementation and get the right internal slot.

Reply:
Since only hosts are able to implement custom module sources, and not user code, this is explicitly designed to ensure `class extends AbstractModuleSource` _doesn't work_.